// web-ifc-viewer/dist/components/display/fills.js
import { BackSide } from 'three';
export class SectionFillManager {
    constructor(IFC, context) {
        this.IFC = IFC;
        this.context = context;
        this.fills = {};
    }
    get(name) {
        return this.fills[name];
    }
    create(name, modelID, ids, material) {
        this.setupMaterial(material);
        const subset = this.getSubset(modelID, ids, material);
        if (!subset)
            return null;
        this.context.scene.addModel(subset);
        this.fills[name] = subset;
        return subset;
    }
    delete(name) {
        const subset = this.fills[name];
        delete this.fills[name];
        this.context.scene.removeModel(subset);
        if (subset.parent)
            subset.removeFromParent();
        subset.geometry.dispose();
    }
    setupMaterial(material) {
        material.clippingPlanes = this.context.getClippingPlanes();
        material.side = BackSide;
        material.polygonOffset = true;
        material.polygonOffsetFactor = -1;
        material.polygonOffsetUnits = 1;
    }
    getSubset(modelID, ids, material) {
        return this.IFC.loader.ifcManager.createSubset({
            modelID,
            ids,
            scene: this.context.getScene(),
            removePrevious: true,
            material
        });
    }
}
